File: Core/Service.cs
using Uuids;

namespace Core
{
    public static class Service
    {
        public static ErrorOr<Handle> Create(string protocolName, string vendorName, string deviceName, Uuid deviceId)
        {
            var handle = Syscalls.ServiceCreate(protocolName, vendorName, deviceName, deviceId);
            if (!handle.IsError()) Process.RegisterHandle(handle.Value());
            return handle;
        }

        public static Optional<Error> Destroy(Handle serviceHandle)
        {
            return new Optional<Error>(Error.NotImplemented);
        }
    }
}

File: rust/chaos-core/src/service.rs
use uuid::Uuid;

use crate::error::Error;
use crate::handle::Handle;
use crate::syscalls;

pub fn create(protocol_name: &str, vendor_name: &str, device_name: &str, device_id: Uuid) -> Result<Handle, Error> {
    syscalls::service_create(protocol_name, vendor_name, device_name, device_id)
}

pub fn connect(protocol_name: &str, vendor_name: Option<&str>, device_name: Option<&str>, device_id: Option<Uuid>) -> Result<Handle, Error> {
    syscalls::service_connect(protocol_name, vendor_name, device_name, device_id)
}